tez-examples/src/main/java/org/apache/tez/examples/Demuxer.java

```diff
@@ -0,0 +1,231 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * Simple example of a demuxer tolerant of data skew problems. The example scans the source files,
+ * partitions records by values, and write them in /path/to/{value}-r-{task id}.
+ */
+public class Demuxer extends TezExampleBase {
+  private static final Logger LOG = LoggerFactory.getLogger(Demuxer.class);
+
+  private static final String DEMUXER_OUTPUT = "demuxerOutput";
+
+  public static void main(String[] args) throws Exception {
+    Demuxer job = new Demuxer();
+    int status = ToolRunner.run(new Configuration(), job, args);
+    System.exit(status);
+  }
+
+  @Override
+  protected void printUsage() {
+    System.err.println("Usage: "
+        + "demuxer <inputPath> <outPath> <numPartitions> [isPrecise(default false)]");
+  }
+
+  @Override
+  protected int validateArgs(String[] otherArgs) {
+    if (otherArgs.length < 3 || otherArgs.length > 4) {
+      return 2;
+    }
+    return 0;
+  }
+
+  @Override
+  protected int runJob(String[] args, TezConfiguration tezConf, TezClient tezClient)
+      throws Exception {
+    String inputDirs = args[0];
+    String outputDir = args[1];
+    int numPartitions = Integer.parseInt(args[2]);
+    if (args.length == 4 && Boolean.parseBoolean(args[3])) {
+      tezConf.set(
+          TezRuntimeConfiguration.TEZ_RUNTIME_REPORT_PARTITION_STATS,
+          ReportPartitionStats.PRECISE.getType());
+    }
+
+    List<Path> inputPaths = Arrays
+        .stream(inputDirs.split(","))
+        .map(Path::new)
+        .collect(Collectors.toList());
+    Path outputPath = new Path(outputDir);
+
+    FileSystem fs = outputPath.getFileSystem(tezConf);
+    outputPath = fs.makeQualified(outputPath);
+    if (fs.exists(outputPath)) {
+      System.err.println("Output directory: " + outputDir + " already exists");
+      return 3;
+    }
+    DAG dag = inputPaths.size() == 1
+        ? createDag(tezConf, inputPaths.get(0), outputPath, numPartitions)
+        : createDagWithUnion(tezConf, inputPaths, outputPath, numPartitions);
+    LOG.info("Running Demuxer");
+    return runDag(dag, isCountersLog(), LOG);
+  }
+
+  private DAG createDag(TezConfiguration tezConf, Path inputPath, Path outputPath,
+      int numPartitions) {
+    Vertex inputVertex = createInputVertex(tezConf, "input", inputPath);
+    Vertex demuxVertex = createDemuxVertex(tezConf, outputPath, numPartitions);
+    OrderedPartitionedKVEdgeConfig edgeConf = OrderedPartitionedKVEdgeConfig
+        .newBuilder(
+            Text.class.getName(),
+            NullWritable.class.getName(),
+            HashPartitioner.class.getName())
+        .setFromConfiguration(tezConf)
+        .build();
+    return DAG
+        .create("Demuxer")
+        .addVertex(inputVertex)
+        .addVertex(demuxVertex)
+        .addEdge(Edge.create(inputVertex, demuxVertex, edgeConf.createDefaultEdgeProperty()));
+  }
+
+  private DAG createDagWithUnion(TezConfiguration tezConf, List<Path> inputPaths, Path outputPath,
```

Comment:
Note: This case doesn't work yet until we merge https://github.com/apache/tez/pull/306

Reply:
https://github.com/apache/tez/pull/306 was closed due to some reasons
is this PR suitable to be adapted to work without that patch?

Reply:
@abstractdog Thanks for asking. I closed #306 and still need to implement an alternative solution. At this point, I removed the case using UNION from the sample.
https://github.com/apache/tez/pull/320/commits/9e0966541383e305d4505f23178aaebfc90181b5